tools/org.eclipse.n4js.hlc.base/src/org/eclipse/n4js/hlc/base/DependenciesHelper.java

```diff
@@ -0,0 +1,169 @@
+		List<File> singleSourceProjectLocations = new ArrayList<>();
+		try {
+			singleSourceProjectLocations.addAll(findProjectsForSingleFiles(srcFiles));
+		} catch (N4JSCompileException e) {
+			System.err.println(Throwables.getStackTraceAsString(e));
+		}
+
+		allProjectsRoots.addAll(discoveredProjectLocations);
+		allProjectsRoots.addAll(singleSourceProjectLocations);
+
+		Map<String, String> dependencies = new HashMap<>();
+
+		DependenciesCollectingUtil.updateMissingDependneciesMap(dependencies,
+				getAvailableProjectsDescriptions(allProjectsRoots));
+
+		return dependencies;
+	}
+
+	private Iterable<ProjectDescription> getAvailableProjectsDescriptions(List<File> allProjectsRoots) {
+		List<ProjectDescription> descriptions = new ArrayList<>();
+		allProjectsRoots.forEach(root -> {
+			File manifest = new File(root, IN4JSProject.N4MF_MANIFEST);
+			if (!manifest.isFile()) {
+				System.out.println("Cannot read manifest at " + root);
+				return;
+			}
+			ProjectDescription projectDescription = ProjectDescriptionProviderUtil.getFromFile(manifest);
+			descriptions.add(projectDescription);
+
+		});
+		return descriptions;
+	}
+
+	/**
+	 * Collects the projects containing the given single source files.
+	 *
+	 * @param sourceFiles
+	 *            the list of single source files
+	 * @return list of N4JS project locations
+	 * @throws N4JSCompileException
+	 *             if no project cannot be found for one of the given files
+	 */
+	// TODO GH-651 copied from
+	// org.eclipse.n4js.generator.headless.N4HeadlessCompiler.findProjectsForSingleFiles(List<File>)
+	private List<File> findProjectsForSingleFiles(List<File> sourceFiles)
+			throws N4JSCompileException {
+
+		Set<URI> result = Sets.newLinkedHashSet();
+
+		for (File sourceFile : sourceFiles) {
+			URI sourceFileURI = URI.createFileURI(sourceFile.toString());
+			URI projectURI = n4jsFileBasedWorkspace.findProjectWith(sourceFileURI);
+			if (projectURI == null) {
+				throw new N4JSCompileException("No project for file '" + sourceFile.toString() + "' found.");
+			}
+			result.add(projectURI);
+		}
+
+		// convert back to Files:
+		return result.stream().map(u -> new File(u.toFileString())).collect(Collectors.toList());
+	}
+
+	/**
+	 * Searches for direct sub-folders containing a File named {@link IN4JSProject#N4MF_MANIFEST}
+	 *
+	 * @param absProjectRoots
+	 *            all project root (must be absolute)
+	 * @return list of directories being a project
+	 */
+	// TODO GH-651 copied org.eclipse.n4js.generator.headless.HeadlessHelper.collectAllProjectPaths(List<File>)
+	static ArrayList<File> collectAllProjectPaths(List<File> absProjectRoots) {
+		ArrayList<File> pDir = new ArrayList<>();
+		for (File projectRoot : absProjectRoots) {
+			Arrays.asList(projectRoot.listFiles(f -> {
+				return f.isDirectory(); // all directrories
+			}))//
+					.stream() //
+					.filter(f -> {
+						File[] list = f.listFiles(f2 -> f2.getName().equals(IN4JSProject.N4MF_MANIFEST));
```

Comment:
my usual comment about understandability here

Reply:
yes, I have refactored it in the GH-651